### ISSUE TITLE
Site page footer links cannot be applied as a template

### ISSUE THREAD

**Issue by diver7** [pnp/pnpframework#76]
### Reporting an Issue or Missing Feature
When you create a link on a the footer of the home page to a page on the site it will only store the portion of the Url after the base Url. So if the footer link Url is "https://tenant.sharepoint.com/sitepages/test.aspx", the footer will only store "/sitepages/test.aspx". When the footer is rendered it will add the base Url back to the link so it works correctly.

If I call "Get-PnPSiteTemplate -Out .\hubfootertemplate.xml -Force -Handlers SiteFooter" to export the footer, the xml file will only have the partial Url to the site page. If I then try to import that footer xml file into another site using "Invoke-PnPSiteTemplate -Path hubfootertemplate.xml" I will get an error stating that the page cannot be opened.

`Invoke-PnPSiteTemplate: C:\test\UpdateFooter_linkerror.ps1:8
Line |
   8 |  Invoke-PnPSiteTemplate -Path hubfootertemplate.xml
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot open "/SitePages/Test.aspx": no such file or folder.`

As a workaround, if I then go back to the original footer and add the base Url to the link again and save it. It will save the entire Url in the footer. Not sure why it does this. Then exporting it and importing as a template will work because the entire Url will be in the exported template xml file. 

### Expected behavior 
Exporting a footer to xml and then importing it to another site should apply the template without an error.

### Actual behavior 
Error occurs:
`Invoke-PnPSiteTemplate: C:\test\UpdateFooter_linkerror.ps1:8
Line |
   8 |  Invoke-PnPSiteTemplate -Path hubfootertemplate.xml
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot open "/SitePages/Test.aspx": no such file or folder.`

### Steps to reproduce behavior
1. Create two communication sites
2. Add a footer to one of the sites with a link to a page on that site.
3. Run this script to apply the footer to the other site: 
```
$hubsite = Read-Host "Please enter the Hub site Url" #https://tenant.sharepoint.com
$global:creds = Get-Credential
Connect-PnPOnline -Url $hubsite -Credentials ($global:creds)
Get-PnPSiteTemplate -Out .\hubfootertemplate.xml -Force -Handlers SiteFooter
Connect-PnPOnline -Url "https://tenant.sharepoint.com/sites/testsite" -Credentials ($global:creds)
Invoke-PnPSiteTemplate -Path hubfootertemplate.xml
```
### What is the version of the Cmdlet module you are running?
0.3.25
PS 7.0.3

### Which operating system/environment are you running PnP PowerShell on?
- [X] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by erwinvanhunen** [comment]
This issue is rather complex. SharePoint stores the footer navigation in the general navigation object available for the site (that contains the top navigation and the quicklaunch nav among others). Objects created in the navigation are either 'external' or 'internal' links. We have no control of how those are being added to the navigation node collection as that is handled by SharePoint itself.

Having said that, when I add a link with fully qualified site and export a template with the sitefooter, the link shows up as:

```xml
 <pnp:FooterLink DisplayName="PnP Test" Url="http://tenant.sharepoint.com{site}/Pages/PnPTest.aspx" />
```

You'll notice the `{site}` token in there. When applying the template we will replace that token with the actual server relative url of the site you're applying the template too. That means, which is the way SharePoint navigation works, that that page -has- to be present in the site. E.g. you'll have to make sure it's in the template if not already on the site. as the server will validate that url. 

If you you want to have a link pointing to a page in the original source site, and the footer in the target site should point to that page in the original site, you will have to modify the template manually or by code unfortunately as there is no way for us to recognize if a link should stay fully qualified or relative. You can modify your template from code in PowerShell too if you want:

```powershell
Connect-PnPOnline -Url https://source
Get-PnPSiteTemplate -Out footer.xml -Handlers SiteFooter
$template = Read-PnPSiteTemplate footer.xml
$link = $template.Footer.FooterLinks | Where-Object{$_.DisplayName -eq "the title of the link"}
$link.Url = "https://yourfullurl"
Save-PnPSiteTemplate -Template $template -Out updatedtemplate.xml
Connect-PnPOnline -Url https://target
Invoke-PnPSiteTemplate -Path ./footer.xml
 ```

Optionally, doing this all in memory without saving the file to XML:

```powershell
Connect-PnPOnline -Url https://source
$template = Get-PnPSiteTemplate -Handlers SiteFooter  -OutputInstance
$link = $template.Footer.FooterLinks | Where-Object{$_.DisplayName -eq "the title of the link"}
$link.Url = "https://yourfullurl"
Connect-PnPOnline -Url https://target
Invoke-PnPSiteTemplate -InputInstance $template
```
In the above script you will notice the $index. That is the zero-bound index of the item in the template. E.g. the 3rd link has an index value of 2.

**Comment by diver7** [comment]
@erwinvanhunen 
I can understand why site navigation would have relative links but I had always thought of the footer as being different from site navigation in that respect although it still makes sense if that's the way it works. My scenario is an intranet with a hub site and many spoke sites. We want to have the same footer on all the spoke sites as we have on the hub site. With all the links being fully qualified. We have some pages on the hub site that the footer links point back to.

What's interesting is the fact that if I create a footer link and paste a fully qualified link (to a page on that site) into the Address field and then save it, SharePoint will change that link to a relative link address. If I edit the same footer link again, I see the relative link. If I then modify the Address field of the link and make it a fully qualified link again and save it, it remains fully qualified. This is how I have been getting my scenario to work. I am not seeing the "{site}" token in the footer area of the template xml as you stated. I do see this token in the navigation area xml of the template though.

So what your telling me is that when I apply the footer template, with a relative link in it, to a site, it will try to validate that the target of that link actually exists on the site. This makes perfect sense. Where this gets a little strange is in the Hub/Spoke site situation. I just tried a test using the SharePoint UI to add a footer link to a spoke site with a relative link address to a page that only exists on the hub site. It let me save the footer link without a validation error and to my amazement the footer link now points to the hub site page when the page is rendered. I think there is still some functionality in the footer source code linking hub and spoke sites in some way.

I think we can consider this issue closed unless my comments above spark some more conversation. I opened it thinking there may be an issue with the way PnP.Powershell was applying the footer template but it doesn't appear that is the case.
